src/jit/lower.cpp

```diff
@@ -3408,6 +3408,9 @@ bool Lowering::IndirsAreEquivalent(GenTreePtr candidate, GenTreePtr storeInd)
     if (pTreeA->OperGet() != pTreeB->OperGet())
         return false;
 
+    if (candidate->gtType != storeInd->gtType)
```

Comment:
@mmitche I had the same concern about this type check being too conservative but I wasn't able to find a repro. I don't see where in your example you have different signedness. I tried to use `ushort` and `short` in the original example but I ended up with TYP_CHAR everywhere.

Anyway, checking the type size should work in this case but there should be an assert that the involved types are integral, otherwise the function might be misused in the future and return true for indirections of type float and int.